### PR TITLE
crimson/osd/osd_operations/client_request: check "can_serve_replica_reads" before getting obc

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/65872
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -400,8 +411,13 @@ ClientRequest::do_process(
   }
 
   if (!obc->obs.exists && !op_info.may_write()) {
-    co_await reply_op_error(pg, -ENOENT);
-    co_return;
+    if (pg->is_primary()) {
+      co_await reply_op_error(pg, -ENOENT);
+      co_return;
+    } else {
+      co_await reply_op_error(pg, -EAGAIN);
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:418]: nit: "enoent on replica osds, returning eagain"

**Review by Matan-B — APPROVED** [review]
We can also assert that the op handled is actually balanced/localized.

File: src/crimson/osd/osd_operations/client_request.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:474]: Done, please take a look, again, thanks.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 2]
```diff
@@ -448,8 +462,18 @@ ClientRequest::do_process(
...
 
   if (!obc->obs.exists && !op_info.may_write()) {
-    co_await reply_op_error(pg, -ENOENT);
-    co_return;
+    if (pg->is_primary()) {
+      co_await reply_op_error(pg, -ENOENT);
+      co_return;
+    } else {
+      DEBUGDPP("{}.{} enoent on replica osds, returning eagain",
+          *pg, *this, this_instance_id);
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:470]: indentation

**Comment by xxhdx1985126** [comment]
> If the replica isn't missing the object and it's not a recent log entry, the correct behavior should be to return ENOENT to the client -- the object would have to actually not exist. The linked bug would have to be due to one of those tests being wrong.

So, I think maybe we should move the `can_serve_replica_reads()` check before the `with_locked_obc()`, so that we can detect whether the object can be served by the replica osd early, and we can always return ENOENT when the object is missing. Am I right?

**Comment by xxhdx1985126** [comment]
> > If the replica isn't missing the object and it's not a recent log entry, the correct behavior should be to return ENOENT to the client -- the object would have to actually not exist. The linked bug would have to be due to one of those tests being wrong.
> 
> So, I think maybe we should move the `can_serve_replica_reads()` check before the `with_locked_obc()`, so that we can detect whether the object can be served by the replica osd early, and we can always return ENOENT when the object is missing. Am I right?

@Matan-B @athanatos I implemented this approach here: https://github.com/ceph/ceph/commit/c590e2e36893dc7f048a7d37262c9a301bc6738a, please take a look.

**Comment by Matan-B** [comment]
> @Matan-B @athanatos I implemented this approach here: [c590e2e](https://github.com/ceph/ceph/commit/c590e2e36893dc7f048a7d37262c9a301bc6738a), please take a look.

This commit is not part of this PR. Is this intentional?

**Comment by xxhdx1985126** [comment]
> > @Matan-B @athanatos I implemented this approach here: [c590e2e](https://github.com/ceph/ceph/commit/c590e2e36893dc7f048a7d37262c9a301bc6738a), please take a look.
> 
> This commit is not part of this PR. Is this intentional?

Um, not really, I didn't know which way is better, so I kept both.

**Comment by xxhdx1985126** [comment]
@athanatos @Matan-B I've just moved that commit here, please take a look again, thanks.

**Review by Matan-B — APPROVED** [review]

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-june-9-C/

**Review by Matan-B — CHANGES_REQUESTED** [review]
This commit made https://tracker.ceph.com/issues/64000 which was resolved in https://github.com/ceph/ceph/pull/55171/files reappear.

We should should probably also move this bit earlier:
```
          if (int res = op_info.set_from_op(&*m, *pg->get_osdmap());
              res != 0) {
	    return reply_op_error(pg, res);
          }
```

See OSD.0: http://qa-proxy.ceph.com/teuthology/matan-2024-06-09_09:52:55-crimson-rados-wip-matanb-crimson-testing-june-9-C-distro-crimson-smithi/7746364/teuthology.log

I'm afraid there are also other failures causes by this, waiting for test to finish:
https://pulpito.ceph.com/matan-2024-06-09_09:52:55-crimson-rados-wip-matanb-crimson-testing-june-9-C-distro-crimson-smithi/

**Comment by xxhdx1985126** [comment]
> This commit made https://tracker.ceph.com/issues/64000 which was resolved in https://github.com/ceph/ceph/pull/55171/files reappear.
> 
> We should should probably also move this bit earlier:
> 
> ```
>           if (int res = op_info.set_from_op(&*m, *pg->get_osdmap());
>               res != 0) {
> 	    return reply_op_error(pg, res);
>           }
> ```
> 
> See OSD.0: http://qa-proxy.ceph.com/teuthology/matan-2024-06-09_09:52:55-crimson-rados-wip-matanb-crimson-testing-june-9-C-distro-crimson-smithi/7746364/teuthology.log
> 
> OSD.0 also: https://pulpito.ceph.com/matan-2024-06-09_09:52:55-crimson-rados-wip-matanb-crimson-testing-june-9-C-distro-crimson-smithi/7746365

Done, please take a look, again.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 3]
```diff
@@ -167,6 +167,30 @@ ClientRequest::interruptible_future<> ClientRequest::with_pg_process_interruptib
       pg.wait_for_active_blocker,
       &decltype(pg.wait_for_active_blocker)::wait));
 
+  if (int res = op_info.set_from_op(&*m, *pg->get_osdmap());
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:170]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/osd_operations/client_request.cc:173:45: error: member reference type 'crimson::osd::PG' is not a pointer; did you mean to use '.'?
  if (int res = op_info.set_from_op(&*m, *pg->get_osdmap());
```

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:170]: Ah, sorry, I'm on a trip now, and the build runs too slow on my notebook, so I directly pushed the code. Will correct it soon.

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc]: Done, please take a look, again, thanks:-)

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-57367

**Pull request merged**